modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductVirtualSettingsFileEntryResourceImpl.java

```diff
@@ -0,0 +1,316 @@
+ */
+@Component(
+	properties = "OSGI-INF/liferay/rest/v1_0/product-virtual-settings-file-entry.properties",
+	scope = ServiceScope.PROTOTYPE,
+	service = ProductVirtualSettingsFileEntryResource.class
+)
+public class ProductVirtualSettingsFileEntryResourceImpl
+	extends BaseProductVirtualSettingsFileEntryResourceImpl {
+
+	@Override
+	public void deleteProductVirtualSettingsFileEntry(Long id)
+		throws Exception {
+
+		CPDVirtualSettingFileEntry cpdVirtualSettingFileEntry =
+			_cpdVirtualSettingFileEntryService.getCPDVirtualSettingFileEntry(
+				id);
+
+		CPDefinitionVirtualSetting cpDefinitionVirtualSetting =
+			cpdVirtualSettingFileEntry.getCPDefinitionVirtualSetting();
+
+		_cpdVirtualSettingFileEntryService.deleteCPDVirtualSettingFileEntry(
+			CPDefinition.class.getName(),
+			cpDefinitionVirtualSetting.getClassPK(), id);
+	}
+
+	@Override
+	public Page<ProductVirtualSettingsFileEntry>
+			getProductVirtualSettingIdProductVirtualSettingsFileEntriesPage(
+				Long id, Pagination pagination)
+		throws Exception {
+
+		CPDefinitionVirtualSetting cpDefinitionVirtualSetting =
+			_cpDefinitionVirtualSettingService.getCPDefinitionVirtualSetting(
+				id);
+
+		return Page.of(
+			TransformUtil.transform(
+				_cpdVirtualSettingFileEntryService.
+					getCPDVirtualSettingFileEntries(
+						CPDefinition.class.getName(),
+						cpDefinitionVirtualSetting.getClassPK(),
+						cpDefinitionVirtualSetting.
+							getCPDefinitionVirtualSettingId(),
+						pagination.getStartPosition(),
+						pagination.getEndPosition()),
+				this::_toProductVirtualSettingsFileEntry),
+			pagination,
+			cpDefinitionVirtualSetting.getCPDVirtualSettingFileEntriesCount());
+	}
+
+	@Override
+	public ProductVirtualSettingsFileEntry getProductVirtualSettingsFileEntry(
+			Long id)
+		throws Exception {
+
+		return _toProductVirtualSettingsFileEntry(
+			_cpdVirtualSettingFileEntryService.getCPDVirtualSettingFileEntry(
+				id));
+	}
+
+	@Override
+	public ProductVirtualSettingsFileEntry patchProductVirtualSettingsFileEntry(
+			Long id, MultipartBody multipartBody)
+		throws Exception {
+
+		ProductVirtualSettingsFileEntry productVirtualSettingsFileEntry =
+			multipartBody.getValueAsNullableInstance(
+				"productVirtualSettingsFileEntry",
+				ProductVirtualSettingsFileEntry.class);
+
+		CPDVirtualSettingFileEntry cpdVirtualSettingFileEntry =
+			_cpdVirtualSettingFileEntryService.getCPDVirtualSettingFileEntry(
+				id);
+
+		CPDefinitionVirtualSetting cpDefinitionVirtualSetting =
+			_cpDefinitionVirtualSettingService.getCPDefinitionVirtualSetting(
+				cpdVirtualSettingFileEntry.getCPDefinitionVirtualSettingId());
+
+		CPDefinition cpDefinition = _cpDefinitionLocalService.getCPDefinition(
+			cpDefinitionVirtualSetting.getClassPK());
+
+		long fileEntryId = cpdVirtualSettingFileEntry.getFileEntryId();
+
+		BinaryFile binaryFile = multipartBody.getBinaryFile("file");
+
+		if (binaryFile != null) {
+			FileEntry fileEntry = FileEntryUtil.addFileEntry(
```

Comment:
probably the FileEntryUtil should be wrapped inside a specific wrapper for this ProductVirtualSettingsFileEntry